app/src/main/java/org/mozilla/fenix/components/metrics/AppStartupTelemetry.kt

```diff
@@ -0,0 +1,109 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.metrics
+
+import android.content.Intent
+import androidx.annotation.VisibleForTesting
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleObserver
+import androidx.lifecycle.OnLifecycleEvent
+import androidx.lifecycle.ProcessLifecycleOwner
+import mozilla.components.support.utils.SafeIntent
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Source
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Type
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Type.COLD
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Type.WARM
+
+/**
+ * Tracks application startup source, type, and whether or not activity has savedInstance to restore
+ * the activity from. Sample metric = [source = COLD, type = APP_ICON, hasSavedInstance = false].
+ * The basic idea is to collect these metrics from different phases of startup through
+ * [AppStartupSourceAndType]and finally report them on Activity's onResume() function.
+ * If application wasn't backgrounded before onResume() is called, we don't report the metric.
+ */
+class AppStartupTelemetry(private val metrics: MetricController) : LifecycleObserver {
+
+    init {
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    private var appOpenedAllSourceStartup = AppStartupSourceAndType()
+
+    private var isMetricRecordedSinceAppWasForegrounded = false
+    private var wasAppCreateCalledBeforeActivityCreate = false
+
+    fun onFenixApplicationOnCreate() {
+        wasAppCreateCalledBeforeActivityCreate = true
+    }
+
+    fun onHomeActivityOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        appOpenedAllSourceStartup.hasSavedInstance = hasSavedInstanceState
```

Comment:
> send a probe saying "error happened" rather than transmit the bad data because we have no way of knowing that we forgot to set one of the vars.

For this reason, I had set up the initial state for all metric fields to be UNKNOWN, so to avoid sending wrong data (random states) if we forget to set a field and set the metric state to UNKNOWN after every time we record the metric.

I changed all the metric fields to be null by default(when init the metric object) for now and added a boolean check to make sure we are never sending a null state. I guess my question for now is, do we crash if we enter the impossible(theoretically) state where a metric is being recorded but some fields are null?

For example, 


    private fun recordMetric() {
        if (!isMetricRecordedSinceAppWasForegrounded && metricDataContainer.isAllDataCollected()) {
            metrics.track(metricDataContainer)
            isMetricRecordedSinceAppWasForegrounded = true
        } else if (!isMetricRecordedSinceAppWasForegrounded && !metricDataContainer.isAllDataCollected()) {
            // we should never be in this state
            //todo do we log it or crash?
            Log.e(Companion.LOG_TAG, "App startup telemetry not sent. Data: $metricDataContainer")
        }
        // we don't want any weird previous states to persist on our next metric record.
        metricDataContainer = AppStartupSourceAndType()
    }

PS: I definitely appreciate all the feedbacks (nits or otherwise) so keep em coming! 😄